### ISSUE TITLE
Errors thrown in Subscriber's next handler should call unsubscribe logic

### ISSUE THREAD

**Issue by blesh** [ReactiveX/RxJS#1186]
It was _supposed_ to be this way in Rx 4.0.6, according to @mattpodwysocki ... but [when I tried it](http://jsbin.com/xuhopiravu/1/edit?html,js,output) it didn't work out that way.

**Comment by blesh** [comment]
Digging into this, I'm not sure this can be implemented reliably. This is because any synchronous call to `next` will happen *before* the subscriber function returns the teardown/unsubscribe logic:

```js
const source = new Observable(observer => {
  // start some expensive async thing
  var id = asyncThing((x => observer.next(x));

  // synchronously emit a value (that will throw ultimately)
  observer.next('wee');

  // return teardown logic for expensive async thing
  return () => {
    cancelAsyncThing(id);
  };
});

source.subscribe((x) => {
  if (x === 'wee') {
    throw new Error('stop having fun!');
  }
});
```

In the above, `cancelAsyncThing` will _never_ be called. It doesn't matter if we wrap the next handler in a try/catch and try to call unsubscribe, because the Observable's setup logic hasn't even completed yet to even _return_ the unsubscribe logic.

There's basically no way to get this to work consistently unless all Observable implementations had a try/finally block.

**I'm not sure what we should do for expected behavior here. But I think it should be consistent so as not to surprise developers.**

cc/ @zenparsing @jhusain @headinthebox

**Comment by blesh** [comment]
I suppose we could squelch all `next`, `error` and `complete` calls after the error is thrown, to allow the subscriber logic to complete and return the teardown, then unsubscribe.

**Comment by trxcllnt** [comment]
@blesh resolved in #1191, which also resolves some of @mattpodwysocki and @zenparsing's concerns about the difference between using callbacks and a duck-typed Observer object.

**Comment by blesh** [comment]
@trxcllnt did you happen to test perf before and after? I suspect that adding several conditionals and a few function calls to each emission is going to impact performance.

**Comment by trxcllnt** [comment]
@blesh I ran `node perf/micro mergemap-observable` and there was no difference. I haven't run the full micro/macro suite though.

**Comment by blesh** [comment]
This is tricky, because the ideal behavior is to halt execution of the subscription logic and teardown only what's been subscribed to at that point.

In effect, it would almost be better to have teardown logic added to `subscriber` via `subscriber.add` than it is to have to get to the end and return the teardown logic.

**Comment by blesh** [comment]
This was resolved by #1191.
